### PR TITLE
[FLINK-16219][runtime] Made AsyncWaitOperator chainable to non-sources.

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

Made AsyncWaitOperator chainable to non-sources.

## Brief change log

- Adding HEAD_AFTER_LEGACY_SOURCE ChainingStrategy.  Operators will be eagerly chained whenever possible, except after legacy sources.
- Made AsyncWaitOperator chainable to non-sources.

## Verifying this change

- Added AsyncWaitOperatorTest#testAsyncOperatorIsNotChainedToSources

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/ChainingStrategy.java [hunk 1]
```diff
@@ -38,16 +38,46 @@
...
+	 * instead of {@link #ALWAYS}.
+	 */
+	HEAD_AFTER_LEGACY_SOURCE {
+		@Override
+		public boolean canBeChainedTo(StreamOperatorFactory<?> headOperator) {
+			return !headOperator.isStreamSource();
+		}
+	};
+
+	public abstract boolean canBeChainedTo(StreamOperatorFactory<?> headOperator);
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/ChainingStrategy.java]: Is it working properly? What if we have `StreamSource -> OP2 -> OP3 -> AsyncWaitOperator`?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperatorTest.java [hunk 2]
```diff
@@ -365,10 +365,10 @@ private void testProcessingTime(AsyncDataStream.OutputMode mode) throws Exceptio
...
 	/**
-	 * Test for the temporary fix to FLINK-13063.
+	 * Tests FLINK-16219.
 	 */
 	@Test
-	public void testAsyncOperatorIsNeverChained() {
+	public void testAsyncOperatorIsNotChainedToSources() {
 		StreamExecutionEnvironment chainEnv = StreamExecutionEnvironment.getExecutionEnvironment();
 
 		DataStream<Integer> input = chainEnv.fromElements(1);
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperatorTest.java]: add a simple mapping function?
```
input = chainEnv.fromElements(1).map(....);
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/ChainingStrategy.java (hunk 1 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/ChainingStrategy.java]: It was not. Good catch.

[the remark this answers is not in the document]
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/ChainingStrategy.java]: Is yielding the root cause of not chainable to legacy sources?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/ChainingStrategy.java [hunk 3]
```diff
@@ -38,16 +38,46 @@
 	 * <p>To optimize performance, it is generally a good practice to allow maximal
 	 * chaining and increase operator parallelism.
 	 */
-	ALWAYS,
+	ALWAYS {
+		@Override
+		public boolean canBeChainedTo(StreamOperatorFactory<?> headOperator) {
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/ChainingStrategy.java]: Renamed vars in a separate hotfix.

For the remainder naming, I think it's the other way around. If I construct head to tail, I'd ask can this node by chained to the head?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 4]
```diff
@@ -602,9 +602,8 @@ public static boolean isChainable(StreamEdge edge, StreamGraph streamGraph) {
 				&& outOperator != null
 				&& headOperator != null
 				&& upStreamVertex.isSameSlotSharingGroup(downStreamVertex)
-				&& outOperator.getChainingStrategy() == ChainingStrategy.ALWAYS
-				&& (headOperator.getChainingStrategy() == ChainingStrategy.HEAD ||
-					headOperator.getChainingStrategy() == ChainingStrategy.ALWAYS)
+				&& outOperator.getChainingStrategy().canBeChainedTo(headOperator)
+				&& headOperator.getChainingStrategy() != ChainingStrategy.NEVER
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java]: A different question imho. First, is upstream supporting chaining at all? Second, is this specific downstream chainable to the upstream.

[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java]: I would prefer to have one method to rule them all (to have one definite answer how to handle the chaining strategy), but as you prefer.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 5]
```diff
@@ -542,14 +542,14 @@ private void connect(Integer headOfChain, StreamEdge edge) {
 
 		physicalEdgesInOrder.add(edge);
 
-		Integer downStreamvertexID = edge.getTargetId();
+		Integer downstreamVertexID = edge.getTargetId();
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java]: nit: `downstream` -> `downStream` & `upstream` -> `upStream`. I quickly git grepped the code, and `downstream` is only used in text/docs. `downStream` is used in the code.

**Comment by AHeise** [comment]
Addressed all comments; let's wait for travis.

**Comment by AHeise** [comment]
@pnowojski green

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 6]
```diff
@@ -587,30 +588,54 @@ private void connect(Integer headOfChain, StreamEdge edge) {
...
-				&& (headOperator.getChainingStrategy() == ChainingStrategy.HEAD ||
-					headOperator.getChainingStrategy() == ChainingStrategy.ALWAYS)
+				&& areOperatorsChainable(upStreamVertex, downStreamVertex, streamGraph)
 				&& (edge.getPartitioner() instanceof ForwardPartitioner)
 				&& edge.getShuffleMode() != ShuffleMode.BATCH
 				&& upStreamVertex.getParallelism() == downStreamVertex.getParallelism()
 				&& streamGraph.isChainingEnabled();
 	}
 
+	private static boolean areOperatorsChainable(
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java]: Can we unit test this method assuming that `StreamNode` and `StreamGraph` are easy to construct/mock?

**Pull request merged**